jax/_src/extend/ffi.py

```diff
@@ -59,3 +67,77 @@ def include_dir() -> str:
   """Get the path to the directory containing header files bundled with jaxlib"""
   jaxlib_dir = os.path.dirname(os.path.abspath(jaxlib.__file__))
   return os.path.join(jaxlib_dir, "include")
+
+
+def ffi_lowering(
+    call_target_name: str,
+    *,
+    operand_layouts: Sequence[Sequence[DimSize]] | None = None,
+    result_layouts: Sequence[Sequence[DimSize]] | None = None,
+    backend_config: Mapping[str, ir.Attribute] | None = None,
+    **lowering_args: Any
+) -> mlir.LoweringRule:
+  """Build a lowering rule for an foreign function interface (FFI) target.
+
+  By default, this lowering rule can use the input and output abstract values to
+  compute the input and output types and shapes for the custom call, assuming
+  row-major layouts.
+
+  If keyword arguments are passed to the lowering rule, these are treated as
+  attributes, and added to `backend_config`.
+
+  Args:
+    call_target_name: The name of the custom call target.
+    operand_layouts: A sequence of layouts (dimension orders) for each operand.
```

Comment:
FWIW I don't think this (layouts) is fully supported today, it's easy to add strides to capture buffer layout, but no one asked for it so far and we don't have any examples

Reply:
@ezhulenev — Oh interesting - good to know! So, if I understand correctly, the custom call always receive buffers in row-major(?) order regardless of the `layout` values passed to mlir?